### PR TITLE
First version of  linear checker

### PR DESCRIPTION
**Pull request by kenarab**
It includes the option --check for running linear checker algorithms which will run only for variables of a type named "Linear".

It includes 3 linearExamples in example path. This examples are included as testcases and are useful for testing and developing

By the moment, neither example is working ok because of rvalue is checked as if it an lvalue.

We should finish  semantics assessment of [linearExample02.con](https://github.com/lambdaclass/concrete/pull/139/files#diff-936e92772c07df222d4be7f7bb914469a06803ff2b6e5439b5ba3fd528483d94) involving consuming a borrowed variable in a function

### CODE REVIEW

File: crates/concrete_check/src/linearity_check.rs [hunk 1]
```diff
@@ -0,0 +1,1086 @@
+use std::collections::HashMap;
+
+use concrete_session::Session;
+
+use self::errors::LinearityError;
+pub mod errors;
+
+use std::path::PathBuf;
+//use concrete_ir::{ProgramBody, FnBody};
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:9]: please remove these commented imports. They can be added afterwards if they are needed.

File: crates/concrete_check/src/linearity_check.rs [hunk 2]
```diff
@@ -0,0 +1,1086 @@
...
+
+    fn _write_once() -> Self {
+        Self::new(0, 1, 0, 0)
+    }
+
+    fn path_once() -> Self {
+        Self::new(0, 0, 0, 1)
+    }
+
+    /* TODO implementation of merge without copying
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:129]: can you summarize here in the TODO what is needed to complete the implementation or why it is commented?

File: crates/concrete_check/src/linearity_check.rs [hunk 3]
```diff
@@ -0,0 +1,1086 @@
...
+
+    fn _merge_list(appearances: Vec<Appearances>) -> Self {
+        appearances
+            .into_iter()
+            .fold(Self::zero(), |acc, x| acc.merge(&x))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct StateTbl {
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:154]: can you add a short doc comment (prefix it with `///` instead of `//`) explaining what is a StateTbl ?

File: crates/concrete_check/src/linearity_check.rs [hunk 4]
```diff
@@ -0,0 +1,1086 @@
...
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+struct LinearityChecker {
+    //state_tbl: StateTbl,
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:268]: why is this commented? can you add a `TODO` with the reason?

File: crates/concrete_check/src/linearity_check.rs [hunk 5]
```diff
@@ -0,0 +1,1086 @@
...
+}
+
+#[allow(dead_code)]
+#[allow(unused_variables)]
+impl LinearityChecker {
+    fn new() -> Self {
+        LinearityChecker {
+            //state_tbl: StateTbl::new(),
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:276]: same here. Add a short comment explaining why it is commented or remove the comment.

File: crates/concrete_check/src/linearity_check.rs [hunk 6]
```diff
@@ -0,0 +1,1086 @@
...
+            "Consuming variable: {} depth {} loop_depth {}",
+            name,
+            depth,
+            loop_depth
+        );
+        if depth == state_tbl.get_loop_depth(name) {
+            state_tbl.update_state(name, &VarState::Consumed);
+            tracing::debug!("Consumed variable: {}", name);
+            /*
+            let mut state = self.state_tbl.get_state(name);
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:297]: can you add a short comment explaining why these lines are commented?

File: crates/concrete_check/src/linearity_check.rs [hunk 7]
```diff
@@ -0,0 +1,1086 @@
...
+                                .count_in_let_statements(name, init)
+                                .merge(&self.count_in_expression(name, cond))
+                                .merge(&self.count_in_assign_statement(name, post))
+                                .merge(&self.count_in_statements(name, block))
+                        }
+                    }
+                }
+                apps
+            }
+            /* Alucination of GPT
```
Comment (juanbono) [inline crates/concrete_check/src/linearity_check.rs:388]: this comment does not belong here. 
We should check the original algorithm as the source of truth

**Review by juanbono — CHANGES_REQUESTED** [review]
See comments

File: crates/concrete_check/src/linearity_check.rs (hunk 7 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:388]: GPT4 infered Statement::Block exists, but it doesn't exists. Remved the comment

File: crates/concrete_check/src/linearity_check.rs (hunk 6 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:297]: Removed the commented code as it was an approach not implemented. Implemented specific error when a variable created outside a loop is consumed inside the loop.

File: crates/concrete_check/src/linearity_check.rs (hunk 5 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:276]: The first approach had a state_tbl object in LinearityChecker. Current implementation generates copies of state_tbl and manage span (table content) with execution flow using recursive calls

File: crates/concrete_check/src/linearity_check.rs (hunk 4 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:268]: Explained in impl LinearityChecker review, just bellow. Removed comment

File: crates/concrete_check/src/linearity_check.rs (hunk 3 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:154]: Done

File: crates/concrete_check/src/linearity_check.rs (hunk 2 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:129]: The approach was an intended optimization, but it is not necessary by now as we are far from optimization stage. Removed the commented code

```
    fn merge(&mut self, other: &Appearances) {
            self.consumed += other.consumed;
            self.write += other.write;
            self.read += other.read;
            self.path += other.path;
    }
```

File: crates/concrete_check/src/linearity_check.rs (hunk 1 above)
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs:9]: Done

**Review by igaray — APPROVED** [review]

File: crates/concrete_check/src/linearity_check.rs [hunk 8]
```diff
@@ -0,0 +1,1024 @@
...
+
+#[derive(Debug, Clone, Copy)]
+struct Appearances {
+    consumed: u32,
+    write: u32,
+    read: u32,
+    path: u32,
+}
+
+// TODO remove. This are structures translated from Austral
```
[the remark this answers is not in the document]
Comment (kenarab) [reply crates/concrete_check/src/linearity_check.rs]: Already removed

**Review by juanbono — APPROVED** [review]
